docker/generate_makefile.py

```diff
@@ -0,0 +1,327 @@
+                                             base_tag=BASE_TAG,
+                                             build_arguments=build_arguments))
+    for benchmark in oss_fuzz_benchmarks:
+        print(
+            OSS_FUZZER_BENCHMARK_TEMPLATE.format(
+                fuzzer=fuzzer,
+                benchmark=benchmark,
+                base_tag=BASE_TAG,
+                build_arguments=build_arguments))
+
+    # Generate rules for building/pulling all target/benchmark pairs.
+    all_benchmarks = benchmarks + oss_fuzz_benchmarks
+    all_build_targets = ' '.join([
+        'build-{0}-{1}'.format(fuzzer, benchmark)
+        for benchmark in all_benchmarks
+    ])
+    all_pull_targets = ' '.join([
+        'pull-{0}-{1}'.format(fuzzer, benchmark) for benchmark in all_benchmarks
+    ])
+    print('build-{fuzzer}-all: {all_targets}'.format(
+        fuzzer=fuzzer, all_targets=all_build_targets))
+    print('pull{fuzzer}-all: {all_targets}'.format(
+        fuzzer=fuzzer, all_targets=all_pull_targets))
+
+
+def main():
+    """Main entry point."""
+    # Output boilerplate used by other templates and generated rules.
+    print(BOILERPLATE)
+
+    # Compute the list of benchmarks. OSS-Fuzz benchmarks are built
+    # differently from standard benchmarks.
+    benchmarks = []
+    oss_fuzz_benchmarks = []
+    for benchmark in os.listdir(BENCHMARKS_DIR):
+        benchmark_path = os.path.join(BENCHMARKS_DIR, benchmark)
+        if not os.path.isdir(benchmark_path):
+            continue
+        if os.path.exists(os.path.join(benchmark_path, 'build.sh')):
+            benchmarks.append(benchmark)
+        if os.path.exists(os.path.join(benchmark_path, 'oss-fuzz.yaml')):
+            oss_fuzz_benchmarks.append(benchmark)
+
+    # Generate the build rules for fuzzer/benchmark pairs.
+    fuzzers_and_variants = []
+    for fuzzer in os.listdir(FUZZERS_DIR):
+        # Skip non-directory files. These do not represent fuzzers.
+        fuzzer_dir = os.path.join(FUZZERS_DIR, fuzzer)
+        if not os.path.isdir(fuzzer_dir):
+            continue
+
+        # The default fuzzer does not contain any special build arguments.
+        generate_fuzzer(fuzzer,
+                        benchmarks,
+                        oss_fuzz_benchmarks,
+                        build_arguments='')
+        fuzzers_and_variants.append(fuzzer)
+
+        # If there are variants, generate rules for them as well.
+        variants_path = os.path.join(fuzzer_dir, 'variants.yaml')
+        if not os.path.exists(variants_path):
+            continue
+
+        with open(variants_path) as file_handle:
+            config = yaml.load(file_handle, yaml.SafeLoader)
+            assert 'variants' in config
+
+        for variant in config['variants']:
+            assert 'name' in variant
+            variant_name = variant['name']
+            # Build arguments are optional for variants.
+            build_arguments = variant.get('build_arguments', [])
```

Comment:
Consider showing an example build argument in the comment to help readers understand what they are.

Reply:
Good idea. Done!